### DISCUSSION TITLE
yet another challenging watermark question

### DISCUSSION THREAD

**Discussion by tugrul** [ImageMagick/ImageMagick#3590] [Digital Image Processing]
Hi,

I have two original images and two watermarked samples that use exact same watermark image like below.

How can I recover original alpha watermark image using this materials? 

I looks pretty doable by formula if I don't miss something.

```
A1 = first original image pixel value
A2 = second original image pixel value
B = watermark pixel value
x = alpha 
C1 = first watermarked image pixel value
C2 = second watermarked image pixel value


A1 * x + B * (1 - x) = C1
A2 * x + B * (1 - x) = C2


B * (1 - x) = C1 - A1 * x
B * (1 - x) = C2 - A2 * x

(C1 - C2) / (A1 - A2) = x
```

materials:

![original_a](https://user-images.githubusercontent.com/163442/115944974-df7bad80-a4c1-11eb-9ef2-d13e3ebfc9e9.png)
![marked_a](https://user-images.githubusercontent.com/163442/115944977-e3a7cb00-a4c1-11eb-82a2-110ab6805790.png)
![original_b](https://user-images.githubusercontent.com/163442/115944980-e6a2bb80-a4c1-11eb-8f3d-fa6815071dd9.png)
![marked_b](https://user-images.githubusercontent.com/163442/115944981-ea364280-a4c1-11eb-9944-3d2574ac3eab.png)

Special thanks to @snibgo for previous challenging question.

**Comment by snibgo** [comment]
Your maths looks dodgy. Is "x" the alpha of a watermark pixel? If so, remembering that alpha=0 is transparent, and alpha=1 means opaque, then:
```
C1 = A1 * (1-x) + B * x
```
... because where the watermark is transparent, we should have x=0 and C1 = A1.

After a similar question last year, I documented more about deriving a semi-transparent watermark from two images. Sadly, my upload process seems to be broken. But I can show the result:

```
call %PICTBAT%findWater2img wm_a1.png wm_a2.png wm_c1.png wm_c2.png wm_calc.png
```
![wm_calc](https://user-images.githubusercontent.com/33812368/115947119-798d2700-a4bd-11eb-97e3-a034c8728f9d.png)
EDIT to add: the script assumes that corresponding pixels of the two un-watermarked images are different in every channel. Where this isn't true, we can't correctly calculate the pixel colour value, nor the pixel alpha. This has happened in your example.

**Comment by snibgo** [comment]
Why don't my uploads work? Because the two locations:
```
http://snibgo.com/im/watermark.htm
http://im.snibgo.com/watermark.htm
```
... used to be the same place, but now they are different places. I don't know why, or how to fix that. Anyhow, try http://snibgo.com/im/watermark.htm#twoimg for now.

EDIT to add: someone must have heard me banging my head on my computer. The new watermark.htm now appears correctly at both locations.

**Comment by tugrul** [comment]
Yes `x` is other half of alpha. You can swap each other. `1 - (C1 - C2) / (A1 - A2) ` gives alpha of transparent image pixels.

Yes there are two exceptions. Alpha 1 (opaque pixels of watermark) and alpha 0 (invisible pixels of watermark) is a problem as you say. This cases break the formula (1/0) and (0/1) problem.

In aside you can use github pages with markdown format to keep your notes well organized. It supports custom domains and you can bind your im.snibgo.com domain to github pages. Also github repo can be open contributions of other people.

**Comment by snibgo** [comment]
alpha=0 or 1 are particular cases, but not exceptions. My script correctly processes these, with no special processing.

alpha=0 shouldn't be a problem. Many pixels in the watermark image will be entirely transparent, so the those parts of the watermark won't change the image at all.

alpha=1 is also not a problem. At pixels where the unwatermarked images are different, but the watermarked images are identical, the watermark pixels must be opaque and equal to either of the watermarked images.
